internal/topology/variables/cluster_variable_defaulting.go

```diff
@@ -30,66 +30,89 @@ import (
-	if len(allErrs) > 0 {
-		return nil, allErrs
+	// Get a map of ClusterVariables and ensure that variables are not defined more than once in Cluster spec.
+	clusterVariablesMap, err := getClusterVariablesMap(clusterVariables)
+	if err != nil {
+		return nil, append(allErrs, field.Invalid(fldPath, clusterVariables,
+			fmt.Sprintf("cluster variables not valid: %s", err)))
 	}
 
+	// Get a map of ClusterClassVariables for each variable name and definition.
+	clusterClassVariablesMap := getClusterClassVariablesMap(clusterClassVariables)
+
 	// allVariables is used to get a full correctly ordered list of variables.
-	allVariables := []string{}
+	allVariables := []clusterv1.ClusterVariable{}
 	// Add any ClusterVariables that already exist.
-	for _, variable := range clusterVariables {
-		allVariables = append(allVariables, variable.Name)
-	}
+	allVariables = append(allVariables, clusterVariables...)
+
 	// Add variables from the ClusterClass, which currently don't exist on the Cluster.
 	for _, variable := range clusterClassVariables {
-		// Continue if the ClusterClass variable already exists.
-		if _, ok := clusterVariablesMap[variable.Name]; ok {
-			continue
+		for _, definition := range variable.Definitions {
+			if _, ok := clusterVariablesMap[variable.Name]; ok {
+				// Continue if the Cluster variable with a definitionFrom this ClusterClass variable exists.
+				if _, ok := clusterVariablesMap[variable.Name][definition.From]; ok {
+					continue
+				}
+				// Continue if the Cluster variable with an empty definitionFrom exists. The user intention here is to
+				// use the default value from a ClusterClass variable with no conflicting variables.
+				if _, ok := clusterVariablesMap[variable.Name][emptyVariableDefinitionFrom]; ok {
+					continue
+				}
+			}
+			allVariables = append(allVariables, clusterv1.ClusterVariable{
+				Name:           variable.Name,
+				DefinitionFrom: definition.From,
+			})
 		}
-
-		allVariables = append(allVariables, variable.Name)
 	}
 
 	// Default all variables.
 	defaultedClusterVariables := []clusterv1.ClusterVariable{}
-	for i, variableName := range allVariables {
-		clusterClassVariable := clusterClassVariablesMap[variableName]
-		clusterVariable := clusterVariablesMap[variableName]
+	for i, variable := range allVariables {
+		clusterClassVariable, found := clusterClassVariablesMap[ccVariableMapKey{name: variable.Name, from: variable.DefinitionFrom}]
+		if !found {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("name"), variable.Name,
+				fmt.Sprintf("variable with name %q from %q is not defined in ClusterClass `status.Variables`.", variable.Name, variable.DefinitionFrom)))
+			continue
+		}
 
-		defaultedClusterVariable, errs := defaultClusterVariable(clusterVariable, clusterClassVariable, fldPath.Index(i), createVariables)
+		// If the variable is defined in the Cluster spec get the definition.
```

Comment:
Didn't this happen in l.86?

Reply:
That's where the ClusterClass variable comes from this is the ClusterVariable